src/app/shared-secret/shared-secret.component.ts

```diff
@@ -0,0 +1,116 @@
+  styleUrls: ['./shared-secret.component.scss']
+})
+export class SharedSecretComponent implements OnInit {
+
+  ownerPublicKeyBase58Check = '';
+  derivedPublicKeyBase58Check = '';
+  messagePublicKeys: string[] = [];
+  errorMsg = '';
+
+  constructor(
+    private accountService: AccountService,
+    private backendApi: BackendAPIService,
+    private cryptoService: CryptoService,
+    public globalVars: GlobalVarsService,
+  ) { }
+
+  ngOnInit(): void {
+    if (!this.globalVars.callback || !this.globalVars.isCallbackValid){
+      this.errorMsg = 'Callback required';
+      return;
+    }
+
+    const params = new URLSearchParams(window.location.search);
+    if (params.get('ownerPublicKey')){
+      this.ownerPublicKeyBase58Check = params.get('ownerPublicKey') as string;
+      if ( !this.accountService.getPublicKeys().includes(this.ownerPublicKeyBase58Check) ){
+        this.errorMsg = 'Owner Public Key not present in browser data';
+        return;
+      }
+    } else {
+      this.errorMsg = 'Owner Public Key required';
+      console.error(this.errorMsg);
+      return;
+    }
+
+    if (params.get('derivedPublicKey')){
+      this.derivedPublicKeyBase58Check = params.get('derivedPublicKey') as string;
+    } else {
+      this.errorMsg = 'Derived Public Key required';
+      console.error(this.errorMsg);
+      return;
+    }
+
+    if (params.get('JWT')){
+      const JWT = params.get('JWT') as string;
+      try {
+        const derivedPublicKey = this.cryptoService.publicKeyToECBuffer(this.derivedPublicKeyBase58Check).toString('hex');
+        const keyEncoder = new KeyEncoder('secp256k1');
+        const encodedPublicKey = keyEncoder.encodePublic(derivedPublicKey, 'raw', 'pem');
+        const decoded = jsonwebtoken.verify(JWT, encodedPublicKey, {algorithms: ['ES256']});
+      } catch (e) {
+        console.error(e);
+        this.errorMsg = 'Failed to verify JWT';
+        return;
+      }
+    } else {
+      this.errorMsg = 'Derived Public Key required';
+      console.error(this.errorMsg);
+      return;
+    }
+
+    if (params.get('messagePublicKeys')){
+      try {
+        this.messagePublicKeys = (params.get('messagePublicKeys') as string).split(',');
+        for (const messageKey of this.messagePublicKeys) {
+          const publicKeyBytes = this.cryptoService.publicKeyToECBuffer(messageKey);
+        }
+      } catch (e) {
+        console.error(e);
+        this.errorMsg = 'Failed to parse Message Public Keys';
+        return;
+      }
+    } else {
+      this.errorMsg = 'Message Public Keys required';
+      console.error(this.errorMsg);
+      return;
+    }
+
+    this.backendApi.GetUserDerivedKeys(this.ownerPublicKeyBase58Check).subscribe( keys => {
+      if (this.derivedPublicKeyBase58Check in keys) {
+        this.backendApi.GetAppState().subscribe( state => {
+          const blockHeight = state.BlockHeight;
+          const key = keys[this.derivedPublicKeyBase58Check];
```

Comment:
Could key be null if the derived key isn't in keys?

Reply:
I think it would fail the check at line #92, or not?

Reply:
I think you're right but can you just Google it and double check? If we get an error here it could cause issues.

Reply:
I'm ready to bet that we will never get to this point and have an undefined/null key; but JUST in case I added an early return.